port/unix/omrsysinfo.c

```diff
@@ -6548,6 +6561,47 @@ isRunningInContainer(struct OMRPortLibrary *portLibrary, BOOLEAN *inContainer)
 		}
 		rc = 0;
 	}
+
+	/* Read the first line in /proc/1/sched if it exists.
+	 * Outside a container, the initialization process is either "init" or "systemd".
+	 * A containerized environment can be assumed for any other initialization process.
+	 */
+	if ((0 == rc) && (!*inContainer) && (0 == access(OMR_PROC_PID_ONE_SCHED_FILE, F_OK))) {
+		FILE *schedFile = fopen(OMR_PROC_PID_ONE_SCHED_FILE, "r");
+		char buffer[PATH_MAX];
+
+		if (NULL == schedFile) {
+			int32_t osErrCode = errno;
+			Trc_PRT_isRunningInContainer_fopen_failed(OMR_PROC_PID_ONE_SCHED_FILE, osErrCode);
+			rc = portLibrary->error_set_last_error(
+					portLibrary,
+					osErrCode,
+					OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_FOPEN_FAILED);
+			goto _end;
+		}
+
+		if (NULL == fgets(buffer, PATH_MAX, schedFile)) {
+			if (0 != ferror(schedFile)) {
+				int32_t osErrCode = errno;
+				Trc_PRT_isRunningInContainer_fgets_failed(OMR_PROC_PID_ONE_SCHED_FILE, osErrCode);
+				rc = portLibrary->error_set_last_error_with_message_format(
+						portLibrary,
+						OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED,
+						"fgets failed to read %s file stream with errno=%d",
+						OMR_PROC_PID_ONE_SCHED_FILE,
+						osErrCode);
+				goto _end;
+			}
+		} else {
+			/* Check if the initialization process is either "init" or "systemd". */
+			if ((NULL == strstr(buffer, "init"))
+			&& (NULL == strstr(buffer, "systemd"))
```

Comment:
There's nothing preventing anyone from creating executables named "init" or "systemd", so this test isn't reliable.

Reply:
If this check is retained, it should check the value is exactly "init" or "systemd", and not just contains these somewhere in the line.

Reply:
> There's nothing preventing anyone from creating executables named "init" or "systemd", so this test isn't reliable.

We are specifically checking `/proc/1/sched` (PID=1). My current understanding is that only the OS can launch the process with PID=1 . In a non-containerized environment, the operating system's init process is either named `systemd` or `init` for PID=1.

> should check the value is exactly "init" or "systemd"

Updated code for an exact check.

Reply:
PID 1 in the container is used for the main program:
```
$ docker run --rm -it centos:7 /bin/head -5 /proc/1/sched
head (1, #threads: 1)
-------------------------------------------------------------------
se.exec_start                                :     174555845.419648
se.vruntime                                  :            26.328459
se.sum_exec_runtime                          :            25.337377
```
As I said, there's nothing stopping me (or anyone) from building an image with as many executables (for whatever purposes) named "init" or "systemd".

Reply:
Yes, this will lead to an incorrect result. I could not find an approach which will work 100% of the time to correctly detect a containerized env. I referred to https://www.baeldung.com/linux/is-process-running-inside-container for the available container detection techniques; but all approaches might fail at some point. I have opened https://github.com/eclipse/omr/issues/6967 to address the drawback of the new approach.